indico/web/client/js/react/components/WTFRemoteSearchDropdown.jsx

```diff
@@ -0,0 +1,280 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2022 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import _ from 'lodash';
+import PropTypes from 'prop-types';
+import React, {useCallback, useEffect, useMemo, useState} from 'react';
+import {Dropdown} from 'semantic-ui-react';
+
+import {Translate, PluralTranslate, Singular, Plural, Param} from 'indico/react/i18n';
+import {indicoAxios, handleAxiosError} from 'indico/utils/axios';
+import {makeAsyncDebounce} from 'indico/utils/debounce';
+
+const highlightSearch = (text, query = '') => {
+  const index = text.toLowerCase().indexOf(query);
+  if (!query || index === -1) {
+    return text;
+  }
+
+  return (
+    <>
+      {text.slice(0, index)}
+      <b>{text.slice(index, index + query.length)}</b>
+      {text.slice(index + query.length)}
+    </>
+  );
+};
+
+/**
+ * Dropdown which can dynamically query the backend
+ * for results using the search input. This is a replacement
+ * for the selectize.js library.
+ *
+ * It can be both controlled and uncontrolled.
+ * To make it controlled pass the value prop to it.
+ */
+export function RemoteSearchDropdown({
+  searchUrl,
+  searchMethod,
+  searchPayload,
+  minTriggerLength,
+  allowById,
+  preload,
+  valueField,
+  labelField,
+  searchField,
+  value: valueFromProps,
+  onChange: onChangeFromProps,
+  dropdownProps,
+}) {
+  const [searchQuery, setSearchQuery] = useState('');
+  const [options, setOptions] = useState([]);
+  const [loading, setLoading] = useState(false);
+
+  const isControlled = valueFromProps !== undefined;
+  const [internalValue, setInternalValue] = useState(null);
+  const value = isControlled ? valueFromProps : internalValue;
+
+  const debounce = useMemo(() => makeAsyncDebounce(200), []);
```

Comment:
I'd keep this outside the component. memoizing works, but elsewhere we just define it outside the function - less code that needs to run during renders for sure ;)